### PR TITLE
Introduce custom timeout when waiting for `aws_ecs_service` to reach a steady state

### LINKED ISSUE #17746: ecs_service with wait_for_steady_state should allow to configure delay and retries
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

when applying `ecs_service` with `wait_for_steady_state=true`, terraform will wait for the service to reach a steady state.
currently there is no way to set the delay and max retries to be used while waiting.
the requested feature is to support passing custom parameters for those values

the default value is 10 minutes:
the function call is here: https://github.com/hashicorp/terraform-provider-aws/blob/master/aws/resource_aws_ecs_service.go#L1201
which defines the default values here: https://github.com/aws/aws-sdk-go/blob/master/service/ecs/waiters.go#L31-L32

### New or Affected Resource(s)

* aws_ecs_service

### PR DESCRIPTION
**Pull request by jackbatzner**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Fixes #17746, #18329, #16012

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSEcsService_withLaunchTypeFargateAndUpdateWaitForSteadyStateWithTimeout'
=== RUN   TestAccAWSEcsService_withLaunchTypeFargateAndUpdateWaitForSteadyStateWithTimeout
=== PAUSE TestAccAWSEcsService_withLaunchTypeFargateAndUpdateWaitForSteadyStateWithTimeout
=== CONT  TestAccAWSEcsService_withLaunchTypeFargateAndUpdateWaitForSteadyStateWithTimeout
--- PASS: TestAccAWSEcsService_withLaunchTypeFargateAndUpdateWaitForSteadyStateWithTimeout (179.35s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       185.647s
...
```

### CODE REVIEW

File: aws/resource_aws_ecs_service.go
Comment (brunhil) [inline aws/resource_aws_ecs_service.go:382]: I'm not 100% sold on this property name and would love suggestions if anyone can come up with a better idea?

Another option I had was to put this in a element block like so,

```
wait_for_steady_state {
   enabled                     = true
   timeout_in_minutes = 15
}
```

**Comment by YakDriver** [comment]
@brunhil Thank you for this PR! I believe this PR would fix #18329. You mention that it partially addresses it. What doesn't it cover?

Since there is a built-in timeout mechanism, we often avoid a custom timeout argument unless it's tied to an API argument. However, there are currently many timeout arguments and likely a good portion are not directly tied to an API argument (e.g., ASG, batch job, codebuild, cloudfront, codedeploy).

It would be great if an eventual solution could address most/all of the related issues:

Related #16012
Related #16184
Related #16526
Related #16888
Related #17118
Related #17746
Related #18329
Related #18868 (this PR)

**Comment by brunhil** [comment]
Thanks for taking a peek @YakDriver ! I've answered a few of your questions below.

> @brunhil Thank you for this PR! I believe this PR would fix #18329. You mention that it partially addresses it. What doesn't it cover?

This actually does fix all of #18329 it appears. I don't known why I put the "partially addresses it" in the first place 😄 . 

> Since there is a built-in timeout mechanism, we often avoid a custom timeout argument unless it's tied to an API argument. However, there are currently many timeout arguments and likely a good portion are not directly tied to an API argument (e.g., ASG, batch job, codebuild, cloudfront, codedeploy).

I agree with this statement. The issue is the built in function has a [timeout of 10 minutes](https://github.com/aws/aws-sdk-go/blob/main/service/ecs/waiters.go#L31-L32). In my experience, AWS can take longer than 10 minutes to fully switch over from the old service to the new one. This often happens when the service is actively being used. 

> It would be great if an eventual solution could address most/all of the related issues:
> 
> Related #16012
> Related #16184
> Related #16526
> Related #16888
> Related #17118
> Related #17746
> Related #18329
> Related #18868 (this PR)

I've updated based on issues you found. Only one more is actually addressed, the others didn't seem too relevant based on my initial understanding of them.

File: aws/resource_aws_ecs_service.go
Comment (nickfaughey) [reply aws/resource_aws_ecs_service.go:382]: 1. Block version is definitely more straightforward and easier to read
2. Most other timing-related parameters (health check timeout, deregistration delay, etc) are in seconds. Probably makes sense to follow that convention here also

**Comment by dr-yd** [comment]
I can confirm the issue with deployments taking longer than 10 minutes - in our case, we require a deregistration delay of 20 minutes which causes the deployment to fail if a long-lasting connection is opened shortly before a new deployment. 

As per AWS documentation, the target is allowed to deregister once it has no more open connections. Thus, from my understanding the 20min are the maximum time Terraform would actually run, but it would be very useful to reflect that in the resource timeout.

The hard-coded 10min limit makes this impossible and the "error" is indistinguishable from an actual failure (like e. g. a missing Docker image). This is a large UX problem if the output should be understood by people who aren't intimately familiar with Terraform.

File: aws/resource_aws_ecs_service.go
Comment (tjhiggins) [reply aws/resource_aws_ecs_service.go:382]: @brunhil Thank you for putting up this pr. Definitely a much needed addition. Could you please update so that it might get merged?

Comment (brunhil) [reply aws/resource_aws_ecs_service.go:382]: I was hoping someone from the providers team would be able to provide some input on this potential config as well? 

Any chance you could provide input @YakDriver / @bflad / @anGie44 ?

**Review by luis-miguel-alves — APPROVED** [review]

**Comment by luis-miguel-alves** [comment]
Could you please merge this? Thank you

**Review by luis-miguel-alves — APPROVED** [review]

**Comment by tylerohlsen** [comment]
@YakDriver @luis-miguel-alves Looks like this PR has been stalled waiting for review.  Is there someone we can ping to get the review process going again?  This feature is something I would really like to use.

**Comment by mkielar** [comment]
I know I'm late to the party, but why introduce a new option when we could just use:
```
timeout {
  create = "10m"
  update = "20m"
}
```
I mean, these two would still have to be implemented, because `aws_ecs_ervice` only supports `delete` timeout at the moment, but the user experience would at least be consistent with other resources and [general conventions](https://www.terraform.io/language/resources/syntax#operation-timeouts).

**Comment by breathingdust** [comment]
Hi all 👋 Just letting you know that this is issue is featured on this quarters [roadmap](https://github.com/hashicorp/terraform-provider-aws/blob/main/ROADMAP.md). If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Comment by gdavison** [comment]
Hi @jackbatzner, thank you for contributing this PR. We think that using the built-in customizable timeouts functionality in Terraform is a better option to address this issue. We've created #25641 instead. Thanks again!

**Pull request closed without merging**
